src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -291,5 +295,203 @@ Status PythonGcsPublisher::PublishFunctionKey(
+    } else if (status.error_code() != grpc::StatusCode::OK) {
+      return Status::Invalid(status.error_message());
+    }
+
+    if (publisher_id_ != reply.publisher_id()) {
+      if (publisher_id_ != "") {
+        RAY_LOG(DEBUG) << "Replied publisher_id " << reply.publisher_id()
+                       << " different from " << publisher_id_
+                       << ", this should only happen"
+                       << " during GCS failover.";
+      }
+      publisher_id_ = reply.publisher_id();
+      max_processed_sequence_id_ = 0;
+    }
+    last_batch_size_ = reply.pub_messages().size();
+    for (auto &message : reply.pub_messages()) {
+      if (message.sequence_id() <= max_processed_sequence_id_) {
+        RAY_LOG(WARNING) << "Ignoring out of order message " << message.sequence_id();
+        continue;
+      }
+      max_processed_sequence_id_ = message.sequence_id();
+      if (message.channel_type() != channel_type_) {
+        RAY_LOG(WARNING) << "Ignoring message from unsubscribed channel "
+                         << message.channel_type();
+        continue;
+      }
+      queue_.emplace_back(std::move(message));
+    }
+  }
+
+  *message = queue_.front();
+  queue_.pop_front();
+
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollError(std::string *key_id,
+                                      int64_t timeout_ms,
+                                      rpc::ErrorTableData *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.error_info_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollLogs(std::string *key_id,
+                                     int64_t timeout_ms,
+                                     rpc::LogBatch *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.log_batch_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollFunctionKey(std::string *key_id,
+                                            int64_t timeout_ms,
+                                            rpc::PythonFunction *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.python_function_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollActor(std::string *key_id,
+                                      int64_t timeout_ms,
+                                      rpc::ActorTableData *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.actor_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::Close() {
+  std::shared_ptr<grpc::ClientContext> current_polling_context;
+  {
+    absl::MutexLock lock(&mu_);
+    closed_ = true;
+    current_polling_context = current_polling_context_;
+  }
+  if (current_polling_context) {
```

Comment:
There is still a data race if you do a shallow copy, and I don't think this stack-allocated variable would outlive the field member anyway.

Reply:
See my above comment, this should now be fixed. The pure purpose of the `current_polling_context` local variable is to keep the underlying context alive (the context itself is thread safe). Copying the shared pointer will achieve that by increasing the reference count.

Reply:
Setting `closed_` to true effectively renders `current_polling_context_` immutable in the current logic (any operation serialized after it will immediately return), so I believe the copy here is unnecessary and we can directly use `current_polling_context_`. It also means there was no data race between this and DoPoll and just between different DoPolls.

I guess the other thing is that only the most recently created polling context will be canceled, any other in-flight polling contexts will not, is this by design?

Reply:
Yes, that was my initial thought here when I wrote the code (that `closed_` acts as a barrier) but I think the way the code is currently written is the clearest and most safe under future refactorings -- it clearly indicates which variables should be protected by the lock through the GUARDED_BY annotation, and the compile checks that we adhere to the promise :)

Since this is long polling, there will only be one poll in effect per instance of this class, so cancelling only the most recently created context is the right behavior (it is the same behavior that the Python code had previously).